jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java

```diff
@@ -217,225 +169,158 @@ public void execute(Runnable task) throws RejectedExecutionException
-                        LOG.debug("{} task={} {}", this, task, ReservedThreadExecutor.this);
-                    if (task != null)
-                        return task;
-
-                    // we have idled out
-                    int size = _count.getLo();
-                    // decrement size if we have not also been stopped.
-                    while (size > 0)
-                    {
-                        if (_count.compareAndSetLo(size, --size))
-                            break;
-                        size = _count.getLo();
-                    }
-                    _state = size >= 0 ? State.IDLE : State.STOPPED;
-                    return STOP;
-                }
-                catch (InterruptedException e)
-                {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("ignored", e);
-                }
-            }
-            _state = State.STOPPED;
-            return STOP;
-        }
-
         @Override
         public void run()
         {
             _thread = Thread.currentThread();
+
             try
             {
-                while (true)
+                while (isRunning())
                 {
-                    long count = _count.get();
+                    int slot = _threads.offer(this);
 
-                    // reduce pending if this thread was pending
-                    int pending = getHi(count) - (_state == State.PENDING ? 1 : 0);
-                    int size = getLo(count);
+                    if (slot < 0)
+                        // no slot available
+                        return;
 
-                    State next;
-                    if (size < 0 || size >= _capacity)
+                    Runnable task = waitForTask();
+                    while (task == null)
                     {
-                        // The executor has stopped or this thread is excess to capacity
-                        next = State.STOPPED;
-                    }
-                    else
-                    {
-                        long now = NanoTime.now();
-                        long lastEmpty = _lastEmptyNanoTime.get();
-                        if (size > 0 && _idleTimeNanos < NanoTime.elapsed(lastEmpty, now) && _lastEmptyNanoTime.compareAndSet(lastEmpty, now))
-                        {
-                            // it has been too long since we hit zero reserved threads, so are "busy" idle
-                            next = State.IDLE;
-                        }
-                        else
+                        if (!isRunning())
+                            return;
+
+                        // shrink if we are already removed or there are other reserved threads.
+                        // There is a small chance multiple threads
+                        // will iterate at the same time and we will hit 0, but that is not a huge problem.
+                        if (getAvailable() >= 2 && _threads.remove(this, slot))
```

Comment:
I would make this "magic" 2 a configurable property, something like `minSize` or `corePoolSize` (how it is called in the JDK).